### DISCUSSION TITLE
Remove undefined input checks from all functions

### DISCUSSION THREAD

**Discussion by aleclarson** [radashi-org/radashi#48] [Ideas]
Currently, the TypeScript definitions don't reflect the current behavior of Radashi functions. In many of the functions, when an input is null or undefined, the function tries to avoid a thrown error, returning whatever makes most sense.

There are 3 possible directions from here.
1. Update the TypeScript definitions to not complain when an input is null or undefined.
2. Remove the falsy input edge cases from the implementation code.
3. Keep it all the way it is.

My argument for taking approach 2 is straight-forward: TypeScript does the job of checking for garbage input already. If you're using a REST API or other untyped data source, you should be using Zod or a similar library to perform runtime type validation. The current behavior of Radashi is **actively harmful**, because it can make it harder to surmise the actual root of an unexpected behavior. “Why is this object empty when I'm expecting it not to be? [Minutes or hours later...] Oh! It's because Radashi is coercing an unexpectedly undefined value into an empty object!”

Alright, that should be enough to begin this discussion. This is open to all opinions. As this is a breaking change, it will be a while before a decision is made.

**Comment by localusercamp** [comment]
Absolutely.
This is the same what we discussed about polymorhpism. The less input types are the better.

**Comment by AQian0** [comment]
Obviously, we need to avoid implicit conversions in functions.
If TypeScript can't restrict types, then why do we use TypeScript?
This helps to reduce the mental burden on the user, who can both clearly understand what kind of arguments a function needs, and not spend a lot of time troubleshooting in the Radashi source code because of undetectable errors caused by implicit conversions.
These changes may be inconvenient in the short term, but the benefits are huge in the long term.

(My English is not very good, so all of the above is from translation software)
